src/cplus_plugin/api/request.py

```diff
@@ -110,125 +170,593 @@ def auth(self):
+        :type scenario_uuid: str
+
+        :return: Cplus API URL for getting scenario detail
+        :rtype: str
+        """
         return f"{self.base_url}/scenario/{scenario_uuid}/detail/"
 
-    def scenario_output_list(self, scenario_uuid):
-        return f"{self.base_url}/scenario_output/{scenario_uuid}/list/?download_all=true&page=1&page_size=100"
+    def scenario_output_list(self, scenario_uuid) -> str:
+        """Cplus API URL for listing scenario output
+
+        :param scenario_uuid: Scenario UUID
+        :type scenario_uuid: str
+
+        :return: Cplus API URL for scenario output list
+        :rtype: str
+        """
+        return (
+            f"{self.base_url}/scenario_output/{scenario_uuid}/"
+            "list/?page=1&page_size=100"
+        )
 
 
 class CplusApiRequest:
-    """Request to Cplus API."""
+    """Class to send request to Cplus API."""
 
     page_size = 50
 
     def __init__(self) -> None:
+        super().__init__()
         self.urls = CplusApiUrl()
+        self.trends_urls = TrendsApiUrl()
+        self._api_token = None
+        self.token_exp = None
+
+    def _get_raw_header_value(self, value: str) -> QtCore.QByteArray:
+        """Get byte array of header name or value.
+
+        :param value: header name/value
+        :type value: str
+
+        :return: bytes array of string value
+        :rtype: QtCore.QByteArray
+        """
+        return QtCore.QByteArray(bytes(value, encoding="utf-8"))
+
+    def _generate_request(self, url: str, headers: dict = {}) -> QNetworkRequest:
+        """Generate request from url and set headers in the request.
+
+        :param url: URL in request
+        :type url: str
+
+        :param headers: header dictionary, defaults to {}
+        :type headers: dict, optional
+
+        :return: request object
+        :rtype: QNetworkRequest
+        """
+        request = QNetworkRequest(QtCore.QUrl(url))
+        self._set_headers(request, headers)
+        return request
+
+    def _set_headers(self, request: QNetworkRequest, headers: dict = {}):
+        """Set headers into a request object.
+
+        :param request: request object
+        :type request: QNetworkRequest
+
+        :param headers: header dictionary, defaults to {}
+        :type headers: dict, optional
+        """
+        for key, value in headers.items():
+            request.setRawHeader(
+                self._get_raw_header_value(key),
+                self._get_raw_header_value(value),
+            )
+
+    def _read_json_response(
+        self, reply: typing.Union[QNetworkReply, QgsNetworkReplyContent]
+    ) -> dict:
+        """Parse json response from reply object.
+
+        :param reply: reply object
+        :type reply: QNetworkReply
+
+        :return: dictionary of the response, empty if failed to parse
+        :rtype: dict
+        """
+        response = {}
+        try:
+            if isinstance(reply, QNetworkReply) or isinstance(reply, MagicMock):
```

Comment:
@zamuzakki is the MagicMock from test? which test is it? I think we can also mock the whole function _read_json_response.